test/issue8/proto.pb.go

```diff
@@ -64,6 +66,11 @@ func (m *Foo) Unmarshal(data []byte) error {
 		}
 		fieldNum := int32(wire >> 3)
 		wireType := int(wire & 0x7)
+
+		if fn := int(fieldNum - 1); fn <= 0 {
```

Comment:
Just as a quick fix, i would generate something like (fn <= 0 && fn >=0), this is somewhat ugly, but should be enough to make it safe.

so the line becomes

p.P(`if fn := int(fieldNum-`, strconv.Itoa(rfMin), `); (fn >= 0 && fn <= `, strconv.Itoa(rfMax-rfMin), `) { `)

To explain what's going on: the code makes a bitmask of all fields, and sets bits to 1 for the fields that are present. Also tries to save on the size of a bitmask, by finding required fields with lowest id and highest id. rfMin/rfMax are those ids. Since array indexes start with 0, the code then adjusts incoming field ids into array offsets + bit offsets.

In hindsight this "optimization" seems unnecessary for 99.9% of uses, so it might be simpler to just remove it, and the code would become much simpler.
Another reasonable suggestion i can make - is move code that sets "this field is present" bits, into specific "case" statements, and generate the code doing it only for required fields and do all the math in advance, this should speed up the parsing code.